libdrgn/dwarf_index.c

```diff
@@ -199,23 +199,44 @@ static inline size_t hash_pair_to_shard(struct hash_pair hp)
 	 */
 	return ((hp.first >>
 		 (8 * sizeof(size_t) - 8 - DRGN_DWARF_INDEX_SHARD_BITS)) &
-		(((size_t)1 << DRGN_DWARF_INDEX_SHARD_BITS) - 1));
+		(((size_t)DRGN_DWARF_INDEX_NUM_SHARDS) - 1));
 }
 
 static void
 drgn_dwarf_index_namespace_init(struct drgn_dwarf_index_namespace *ns,
 				struct drgn_dwarf_index *dindex)
 {
-	array_for_each(shard, ns->shards) {
-		omp_init_lock(&shard->lock);
-		drgn_dwarf_index_die_map_init(&shard->map);
-		drgn_dwarf_index_die_vector_init(&shard->dies);
-	}
+	ns->shards = NULL;
 	ns->dindex = dindex;
 	drgn_dwarf_index_pending_die_vector_init(&ns->pending_dies);
 	ns->saved_err = NULL;
 }
 
+static struct drgn_error *
+drgn_dwarf_index_namespace_shards_init(struct drgn_dwarf_index_namespace *ns)
+{
+	// Check once for init before acquiring to avoid lock contention once things are initialized.
+	if (ns->shards)
+		return NULL;
+	#pragma omp critical(drgn_dwarf_index_namespace_shards_init)
+	{
+		if (!ns->shards) {
+			struct drgn_dwarf_index_shard* shards = malloc_array(
+				DRGN_DWARF_INDEX_NUM_SHARDS, sizeof(*ns->shards));
+			for (size_t i = 0; shards && i < DRGN_DWARF_INDEX_NUM_SHARDS; i++) {
```

Comment:
I'd prefer an explicit `if (!shards) return ...;` check rather than putting it in the loop condition.